src/composer.ts

```diff
@@ -232,33 +335,108 @@ export class AtomicTransactionComposer {
         );
       }
 
-      appArgTypes.push(argSpec.type);
-      appArgValues.push(argValue);
+      if (abiTypeIsReference(argType)) {
+        refArgIndexToBasicArgIndex.set(
+          refArgTypes.length,
+          basicArgTypes.length
+        );
+        refArgTypes.push(argType);
+        refArgValues.push(argValue);
+        // treat the reference as a uint8 for encoding purposes
+        argType = new ABIUintType(8);
+      }
+
+      if (typeof argType === 'string') {
+        throw new Error(`Unknown ABI type: ${argType}`);
+      }
+
+      basicArgTypes.push(argType);
+      basicArgValues.push(argValue);
+    }
+
+    const resolvedRefIndexes: number[] = [];
+    const foreignAccounts: string[] = [];
+    const foreignApps: number[] = [];
+    const foreignAssets: number[] = [];
+    for (let i = 0; i < refArgTypes.length; i++) {
+      const refType = refArgTypes[i];
+      const refValue = refArgValues[i];
+      let resolved = 0;
+
+      switch (refType) {
+        case ABIReferenceType.account: {
+          const addressType = new ABIAddressType();
+          const address = addressType.decode(addressType.encode(refValue));
+          resolved = populateForeignArray(address, foreignAccounts, sender);
+          break;
+        }
+        case ABIReferenceType.application: {
+          const uint64Type = new ABIUintType(64);
+          const refAppID = uint64Type.decode(uint64Type.encode(refValue));
+          if (refAppID > Number.MAX_SAFE_INTEGER) {
```

Comment:
Could this check be moved into `populateForeignArray` and have that method take `ABIType` as a parameter as well?

Reply:
This is a good suggestion, I'll attempt this change now

Reply:
I tried implementing this, but the change would need to have much more complicated type annotations, since:
* TypeScript knows that specific types like `ABIUintType` and `ABIAddressType` return `bigint` and `string` respectively, which makes type checking easier. But if we pass a generic `ABIType` instance to the function, then we lose information about the return type and can only assume it returns `ABIValue`.
* `ABIUintType` actually returns a `bigint`, but we invoke `populateForeignArray` with its generic `Type` set to `number` since the foreign ASA and app arrays need to be `number[]`. This causes a problem because we need to add extra code to convert the `bigint` to a `number` that should only run in this case.

I don't think the extra complexity is worth it in this case.